File: q0988/q0988_dfs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private String smallest;

    public String smallestFromLeaf(TreeNode root) {
        smallest = "";
        dfs(root, "");
        return smallest;
    }

    private void dfs(TreeNode node, String currString) {
        if (node == null) return;

        currString = (char)(node.val + 'a') + currString;
        if (node.left == null && node.right == null) {
            if (smallest.length() == 0 || smallest.compareTo(currString) > 0) {
                smallest = currString;
            }
        }

        dfs(node.left, currString);
        dfs(node.right, currString);
    }
}

File: q0988/q0988_dfs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        self.smallest = ""
        self.dfs(root, "")
        return self.smallest
        
    def dfs(self, node, currString):
        if not node:
            return
        
        currString = chr(ord('a') + node.val) + currString
        if not node.left and not node.right:
            if self.smallest == "" or self.smallest > currString:
                self.smallest = currString
            return
        
        self.dfs(node.left, currString)
        self.dfs(node.right, currString)